fmilibcpp is not installed

It seems fmilibcpp is never installed. It is necessary for linking with libecos (along with fmi4c and simple_socket, which are installed). I'm writing a test to evaluate the performance of libecos. Here's my code so far:

ecosperftest.cpp

```C++
#include <string>
#include "ecos/structure/simulation_structure.hpp"

using namespace ecos;
using namespace std;

int main() {
    simulation_structure ss;
    ss.add_model(string("foo"), string("foo.fmu"));
    return 0;
}
```

CMakeLists.txt

```cmake
project(ecosperftest)

cmake_minimum_required(VERSION 3.20.5)

set(CMAKE_CXX_COMPILER g++-13)
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fexceptions -std=c++20")
include_directories("$ENV{HOME}/local/include")
link_directories("$ENV{HOME}/local/lib")

add_executable(ecosperftest ecosperftest.cpp)
target_link_libraries(ecosperftest ecos tbb fmilibcpp fmi4c simple_socket)
```

I manged to get libfmilibcpp to install by applying the following patch:

```diff
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index e5edb0b..e50e122 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -117,22 +117,7 @@ configure_file(
         @ONLY
 )
 
-add_library(fmilibcpp OBJECT
-
-        "fmilibcpp/buffered_slave.hpp"
-        "fmilibcpp/fmicontext.hpp"
-        "fmilibcpp/fmu.hpp"
-        "fmilibcpp/model_description.hpp"
-        "fmilibcpp/scalar_variable.hpp"
-        "fmilibcpp/slave.hpp"
-
-        "fmilibcpp/fmi1/fmi1_fmu.hpp"
-        "fmilibcpp/fmi1/fmi1_slave.hpp"
-        "fmilibcpp/fmi1/fmi1_model_description.hpp"
-
-        "fmilibcpp/fmi2/fmi2_fmu.hpp"
-        "fmilibcpp/fmi2/fmi2_slave.hpp"
-        "fmilibcpp/fmi2/fmi2_model_description.hpp"
+add_library(fmilibcpp
 
         "fmilibcpp/fmu.cpp"
 
@@ -150,6 +135,7 @@ add_library(fmilibcpp OBJECT
 
         "external/pugixml/pugixml.cpp"
 )
+install(TARGETS fmilibcpp)
 set_target_properties(fmilibcpp PROPERTIES POSITION_INDEPENDENT_CODE ON)
 target_link_libraries(fmilibcpp PRIVATE fmi4c)
 target_include_directories(fmilibcpp
```

Removing the .hpp files is maybe superfluous. Anyway, with this change I'm able to link against libfmilibcpp and run the program.

Because bookworm ships with g++-6 which lacks C++20 support, g++-13 must be manually specified. Doing so while adding ecos using FetchContent makes CMake angry:

```cmake
set(CMAKE_CXX_COMPILER g++-13)
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fexceptions -std=c++20")

include(FetchContent)
set(ECOS_BUILD_CLI OFF)
set(ECOS_BUILD_CLIB OFF)    # Set to ON for building C API
set(ECOS_WITH_PROXYFMU OFF) # Set to ON for remoting
FetchContent_Declare(
        ecos
        GIT_REPOSITORY https://github.com/Ecos-platform/ecos.git
        GIT_TAG 4fff699d7ebe817ff7f56ad39b81c0061c809975
)
FetchContent_MakeAvailable(ecos)
```

```
  The CMAKE_CXX_COMPILER:

    g++-13

  is not a full path and was not found in the PATH.

  Tell CMake where to find the compiler by setting either the environment
  variable "CXX" or the CMake cache entry CMAKE_CXX_COMPILER to the full path
  to the compiler, or to the compiler name if it is in the PATH.
```

Having to give CMake the full path to the compiler is typically not necessary and is highly annoying. I don't know why this happens here.

Using ExternalProject instead of FetchContent() makes compilation possible:

```cmake
set(ECOS_INSTALL_DIR "${CMAKE_CURRENT_BINARY_DIR}/ecos-install")

include(ExternalProject)
ExternalProject_Add(ecos-ext
    GIT_REPOSITORY https://github.com/Ecos-platform/ecos.git
    GIT_TAG 4fff699d7ebe817ff7f56ad39b81c0061c809975
    CMAKE_ARGS "-DECOS_BUILD_CLI:BOOL=OFF;-DECOS_BUILD_CLIB:BOOL=OFF;-DECOS_WITH_PROXYFMU:BOOL=OFF;-DCMAKE_CXX_COMPILER=g++-13;-DCMAKE_INSTALL_PREFIX=${ECOS_INSTALL_DIR}"
)

include_directories("${ECOS_INSTALL_DIR}/include")
link_directories("${ECOS_INSTALL_DIR}/lib")

add_executable(ecosperftest ecosperftest.cpp)
target_link_libraries(ecosperftest ecos fmi4c tbb)
add_dependencies(ecosperftest ecos-ext)
```

However, the same issue with libfmilibcpp not being installed and its symbols being missing from libecos persists.

```
file_model_sub_resolver.cpp:(.text._ZN4ecos9fmi_modelC2ERKNSt10filesystem7__cxx114pathEb[_ZN4ecos9fmi_modelC5ERKNSt10filesystem7__cxx114pathEb]+0x4b): undefined reference to `fmilibcpp::loadFmu(std::filesystem::__cxx11::path const&, bool)'
```

Also not every project uses CMake so requiring users to also use CMake is not a good idea.

#39 doesn't compile:

```
FAILED: src/CMakeFiles/proxy_fmu.dir/proxyfmu/proxy_slave.cpp.o 
/usr/lib/ccache/g++-13  -I/home/thardin/biz/iter/fmifast/build/src/ecosperftest/ecos-ext-prefix/src/ecos-ext/src/external/spdlog -I/home/thardin/biz/iter/fmifast/build/src/ecosperftest/ecos-ext-prefix/src/ecos-ext/include -I/home/thardin/biz/iter/fmifast/build/src/ecosperftest/ecos-ext-prefix/src/ecos-ext/src -I/home/thardin/biz/iter/fmifast/build/src/ecosperftest/ecos-ext-prefix/src/ecos-ext-build/thirdparty -I/home/thardin/biz/iter/fmifast/build/src/ecosperftest/ecos-ext-prefix/src/ecos-ext/src/external/flatbuffers -O3 -fPIC -std=gnu++20 -MD -MT src/CMakeFiles/proxy_fmu.dir/proxyfmu/proxy_slave.cpp.o -MF src/CMakeFiles/proxy_fmu.dir/proxyfmu/proxy_slave.cpp.o.d -o src/CMakeFiles/proxy_fmu.dir/proxyfmu/proxy_slave.cpp.o -c /home/thardin/biz/iter/fmifast/build/src/ecosperftest/ecos-ext-prefix/src/ecos-ext/src/proxyfmu/proxy_slave.cpp
In file included from /home/thardin/biz/iter/fmifast/build/src/ecosperftest/ecos-ext-prefix/src/ecos-ext/src/proxyfmu/proxy_slave.cpp:2:
/home/thardin/biz/iter/fmifast/build/src/ecosperftest/ecos-ext-prefix/src/ecos-ext/src/proxyfmu/proxy_slave.hpp:8:10: fatal error: simple_socket/SimpleConnection.hpp: No such file or directory
    8 | #include <simple_socket/SimpleConnection.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -187,7 +187,9 @@ if (ECOS_WITH_PROXYFMU)
 
     target_compile_definitions(libecos PRIVATE "ECOS_WITH_PROXYFMU")
 
-    target_link_libraries(libecos INTERFACE ws2_32)
+    if (WIN32)
+        target_link_libraries(libecos INTERFACE ws2_32)
+    endif ()
 
     target_include_directories(libecos
             PRIVATE

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -156,7 +156,7 @@ target_include_directories(fmilibcpp
         "${PROJECT_SOURCE_DIR}/include"
         "${PROJECT_SOURCE_DIR}/src")
 
-add_library(libecos ${publicHeadersFull} ${privateHeaders} ${sources} ${generatedFiles})
+add_library(libecos ${publicHeadersFull} ${privateHeaders} ${sources} ${generatedFiles} "$<TARGET_OBJECTS:fmi4c>" "$<TARGET_OBJECTS:fmilibcpp>")
 set_target_properties(libecos PROPERTIES PREFIX "")
 set_target_properties(libecos PROPERTIES POSITION_INDEPENDENT_CODE ON)
 target_compile_features(libecos PUBLIC "cxx_std_20")
@@ -171,11 +171,7 @@ target_include_directories(libecos
         "${THIRD_PARTY_INCLUDE_DIR}"
         "${CMAKE_CURRENT_SOURCE_DIR}"
 )
-target_link_libraries(libecos
-        PRIVATE
-        fmi4c
-        "$<TARGET_OBJECTS:fmilibcpp>"
-)
+
 if (UNIX)
     target_link_libraries(libecos INTERFACE stdc++fs tbb)
 endif ()

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -50,22 +50,6 @@ set(privateHeaders
 
 )
 
-if (ECOS_WITH_PROXYFMU)
-    list(APPEND privateHeaders
-
-            "ecos/fmi/proxy/proxy_model.hpp"
-
-            "ecos/resolvers/proxy_model_sub_resolver.hpp"
-
-            "proxyfmu/opcodes.hpp"
-            "proxyfmu/process_helper.hpp"
-            "proxyfmu/proxy_fmu.hpp"
-            "proxyfmu/proxy_slave.hpp"
-            "proxyfmu/remote_info.hpp"
-
-    )
-endif ()
-
 set(sources
 
         "ecos/model_resolver.cpp"
@@ -93,18 +77,6 @@ set(sources
         "ecos/util/plotter.cpp"
 
 )
-
-if (ECOS_WITH_PROXYFMU)
-    list(APPEND sources
-
-            "ecos/resolvers/proxy_model_sub_resolver.cpp"
-
-            "proxyfmu/proxy_fmu.cpp"
-            "proxyfmu/proxy_slave.cpp"
-
-            "external/flatbuffers/flatbuffers/util.cpp"
-    )
-endif ()
 
 # Generate sources from .in files
 set(generatedFiles "${GENERATED_SRC_DIR}/ecos/lib_info.cpp")
@@ -156,7 +128,42 @@ target_include_directories(fmilibcpp
         "${PROJECT_SOURCE_DIR}/include"
         "${PROJECT_SOURCE_DIR}/src")
 
-add_library(libecos ${publicHeadersFull} ${privateHeaders} ${sources} ${generatedFiles} "$<TARGET_OBJECTS:fmi4c>" "$<TARGET_OBJECTS:fmilibcpp>")
+add_library(proxy_fmu OBJECT
+
+        "ecos/fmi/proxy/proxy_model.hpp"
+
+        "ecos/resolvers/proxy_model_sub_resolver.hpp"
+
+        "proxyfmu/opcodes.hpp"
+        "proxyfmu/process_helper.hpp"
+        "proxyfmu/proxy_fmu.hpp"
+        "proxyfmu/proxy_slave.hpp"
+        "proxyfmu/remote_info.hpp"
+
+        "ecos/resolvers/proxy_model_sub_resolver.cpp"
+
+        "proxyfmu/proxy_fmu.cpp"
+        "proxyfmu/proxy_slave.cpp"
+
+        "external/flatbuffers/flatbuffers/util.cpp"
+)
+target_link_libraries(proxy_fmu PRIVATE simple_socket)
+set_target_properties(proxy_fmu PROPERTIES POSITION_INDEPENDENT_CODE ON)
+target_include_directories(proxy_fmu
+        PUBLIC
+        "${CMAKE_CURRENT_SOURCE_DIR}/external/spdlog"
+        PRIVATE
+        "${PROJECT_SOURCE_DIR}/include"
+        "${PROJECT_SOURCE_DIR}/src"
+        "${THIRD_PARTY_INCLUDE_DIR}"
+        "${CMAKE_CURRENT_SOURCE_DIR}/external/flatbuffers"
+)
+
+set(all_sources ${publicHeadersFull} ${privateHeaders} ${sources} ${generatedFiles} "$<TARGET_OBJECTS:fmi4c>" "$<TARGET_OBJECTS:fmilibcpp>")
+if (ECOS_WITH_PROXYFMU)
+    list(APPEND all_sources "$<TARGET_OBJECTS:simple_socket>" "$<TARGET_OBJECTS:proxy_fmu>")
+endif ()
+add_library(libecos ${all_sources})
 set_target_properties(libecos PROPERTIES PREFIX "")
 set_target_properties(libecos PROPERTIES POSITION_INDEPENDENT_CODE ON)
 target_compile_features(libecos PUBLIC "cxx_std_20")
@@ -180,14 +187,11 @@ if (ECOS_WITH_PROXYFMU)
 
     target_compile_definitions(libecos PRIVATE "ECOS_WITH_PROXYFMU")
 
+    target_link_libraries(libecos INTERFACE ws2_32)
+
     target_include_directories(libecos
             PRIVATE
             "${CMAKE_CURRENT_SOURCE_DIR}/external/flatbuffers"
-    )
-
-    target_link_libraries(libecos
-            PRIVATE
-            simple_socket
     )
 
     # Make sure executable is up to date

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -128,36 +128,38 @@ target_include_directories(fmilibcpp
         "${PROJECT_SOURCE_DIR}/include"
         "${PROJECT_SOURCE_DIR}/src")
 
-add_library(proxy_fmu OBJECT
+if (ECOS_WITH_PROXYFMU)
+    add_library(proxy_fmu OBJECT
 
-        "ecos/fmi/proxy/proxy_model.hpp"
+            "ecos/fmi/proxy/proxy_model.hpp"
 
-        "ecos/resolvers/proxy_model_sub_resolver.hpp"
+            "ecos/resolvers/proxy_model_sub_resolver.hpp"
 
-        "proxyfmu/opcodes.hpp"
-        "proxyfmu/process_helper.hpp"
-        "proxyfmu/proxy_fmu.hpp"
-        "proxyfmu/proxy_slave.hpp"
-        "proxyfmu/remote_info.hpp"
+            "proxyfmu/opcodes.hpp"
+            "proxyfmu/process_helper.hpp"
+            "proxyfmu/proxy_fmu.hpp"
+            "proxyfmu/proxy_slave.hpp"
+            "proxyfmu/remote_info.hpp"
 
-        "ecos/resolvers/proxy_model_sub_resolver.cpp"
+            "ecos/resolvers/proxy_model_sub_resolver.cpp"
 
-        "proxyfmu/proxy_fmu.cpp"
-        "proxyfmu/proxy_slave.cpp"
+            "proxyfmu/proxy_fmu.cpp"
+            "proxyfmu/proxy_slave.cpp"
 
-        "external/flatbuffers/flatbuffers/util.cpp"
-)
-target_link_libraries(proxy_fmu PRIVATE simple_socket)
-set_target_properties(proxy_fmu PROPERTIES POSITION_INDEPENDENT_CODE ON)
-target_include_directories(proxy_fmu
-        PUBLIC
-        "${CMAKE_CURRENT_SOURCE_DIR}/external/spdlog"
-        PRIVATE
-        "${PROJECT_SOURCE_DIR}/include"
-        "${PROJECT_SOURCE_DIR}/src"
-        "${THIRD_PARTY_INCLUDE_DIR}"
-        "${CMAKE_CURRENT_SOURCE_DIR}/external/flatbuffers"
-)
+            "external/flatbuffers/flatbuffers/util.cpp"
+    )
+    target_link_libraries(proxy_fmu PRIVATE simple_socket)
+    set_target_properties(proxy_fmu PROPERTIES POSITION_INDEPENDENT_CODE ON)
+    target_include_directories(proxy_fmu
+            PUBLIC
+            "${CMAKE_CURRENT_SOURCE_DIR}/external/spdlog"
+            PRIVATE
+            "${PROJECT_SOURCE_DIR}/include"
+            "${PROJECT_SOURCE_DIR}/src"
+            "${THIRD_PARTY_INCLUDE_DIR}"
+            "${CMAKE_CURRENT_SOURCE_DIR}/external/flatbuffers"
+    )
+endif ()
 
 set(all_sources ${publicHeadersFull} ${privateHeaders} ${sources} ${generatedFiles} "$<TARGET_OBJECTS:fmi4c>" "$<TARGET_OBJECTS:fmilibcpp>")
 if (ECOS_WITH_PROXYFMU)

